### PR TITLE
fix: sql lab refetch button

### PR DESCRIPTION
**Pull request by graceguo-supercat**
### SUMMARY
See #15109, SQL Lab show "Refetch Results" button while fetching large query results. The fix in #15109 works, but it introduce another issue so we had to revert it. This PR adds extra fix on top of it.

See the 2nd commit: in airbnb most users still using client side localStorage to store query history. it seems sometimes the queries data was broken or messed up (for example, after user clear localStorage or delete some queries from History Tab). When user tries to re-run a query and sql lab start polling, sql lab can not find query id in the Redux state, which will cause exception, and show offline status. 

### ADDITIONAL INFORMATION

- [x] Has associated issue: #15109
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/src/SqlLab/reducers/sqlLab.js [hunk 1]
```diff
@@ -520,7 +520,20 @@ export default function sqlLabReducer(state = {}, action) {
...
+            ...state.queries[id],
+            ...changedQuery,
+            // race condition:
+            // because of async behavior, sql lab may still poll a couple of seconds
+            // when it started fetching or finished rendering results
+            state:
+              currentState === 'success' &&
+              ['fetching', 'success'].includes(prevState)
+                ? prevState
+                : currentState,
```
Comment (ktmud) [inline superset-frontend/src/SqlLab/reducers/sqlLab.js:535]: Could this be simplified to `state: currentState == 'fetching' && prevState === 'success' ? 'success' : currentState`?

**Review by ktmud — APPROVED** [review]
LGTM given #15109  was already approved

**Comment by ktmud** [comment]
Reading the original description of #15109, I'm also wondering whether we can just replace the first "success" state to "fetching"?

**Comment by graceguo-supercat** [comment]
> Reading the original description of #15109, I'm also wondering whether we can just replace the first "success" state to "fetching"?
I think what is this PR doing. 
client-side can set `fetching` state, but engine-side, the query state in query table, is `success`. In the next polling, API will return `status: success` and try to overwrite redux state.

**Comment by ktmud** [comment]
I mean where is the original first `success` state set? Does it have to be `success` for it to not break something else?

**Comment by graceguo-supercat** [comment]
> I mean where is the original first `success` state set? Does it have to be `success` for it to not break something else?

the first success is read from polling API, which is stored in queries table.

**Comment by ktmud** [comment]
Got it. Thanks for the clarification!

**Pull request merged**